{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Definitions"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-02-01T16:31:16.758240Z",
     "start_time": "2025-02-01T16:31:13.240728Z"
    }
   },
   "source": [
    "import pandas as pd\n",
    "\n",
    "def fasta_to_dataframe(fasta_file):\n",
    "    headers = []\n",
    "    sequences = []\n",
    "    with open(fasta_file, 'r') as f:\n",
    "        current_sequence = []\n",
    "        current_header = None\n",
    "        for line in f:\n",
    "            line = line.strip()\n",
    "            if line.startswith('>'):  # Header line\n",
    "                if current_header is not None:\n",
    "                    # Save the previous sequence\n",
    "                    headers.append(current_header)\n",
    "                    sequences.append(''.join(current_sequence))\n",
    "                current_header = line[1:]  # Remove '>'\n",
    "                current_sequence = []\n",
    "            else:\n",
    "                current_sequence.append(line)\n",
    "        # Add the last sequence\n",
    "        if current_header is not None:\n",
    "            headers.append(current_header)\n",
    "            sequences.append(''.join(current_sequence))\n",
    "    # Create a DataFrame\n",
    "    return pd.DataFrame({'Header': headers, 'Sequence': sequences})"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "source": "## Read and Format FASTA file",
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "fasta_file = \"../../data/DB.COX1.trimmed.fna\"\n",
    "cutadapt_result = fasta_to_dataframe(fasta_file)\n",
    "\n",
    "cutadapt_result[['BOLD Metadata', 'Taxonomy']] = cutadapt_result['Header'].str.split(';', n=1, expand=True)\n",
    "cutadapt_result[['BOLD ID', 'Read ID', 'Country']] = cutadapt_result['BOLD Metadata'].str.split('|', n=2, expand=True)\n",
    "cutadapt_result['BOLD ID'] = cutadapt_result['BOLD ID'].str.removeprefix('BOLD:')\n",
    "\n",
    "tax_columns = ['Phylum', 'Class', 'Order', 'Family', 'Genus', 'Species']\n",
    "cutadapt_result[tax_columns] = cutadapt_result['Taxonomy'].str.extract(\n",
    "    r'p:([^,]+),c:([^,]+),o:([^,]+),f:([^,]+),g:([^,]+),s:([^;]+)'\n",
    ")\n",
    "\n",
    "cutadapt_result = cutadapt_result.drop(columns=['Header', 'BOLD Metadata', 'Taxonomy'])\n",
    "print(cutadapt_result.head())"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-02-01T16:31:38.592427Z",
     "start_time": "2025-02-01T16:31:16.774130Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                            Sequence  BOLD ID       Read ID  \\\n",
      "0  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...  AAA7085  BLPDA1021-18   \n",
      "1  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...  AAA7085  MHMYN6354-14   \n",
      "2  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...  AAA7085   MHASB734-07   \n",
      "3  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...  AAA7085   MHATB310-06   \n",
      "4  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...  AAA7085  BLPAA6663-17   \n",
      "\n",
      "      Country      Phylum    Class        Order       Family    Genus  \\\n",
      "0  Costa_Rica  Arthropoda  Insecta  Lepidoptera  Saturniidae  Lonomia   \n",
      "1  Costa_Rica  Arthropoda  Insecta  Lepidoptera  Saturniidae  Lonomia   \n",
      "2  Costa_Rica  Arthropoda  Insecta  Lepidoptera  Saturniidae  Lonomia   \n",
      "3  Costa_Rica  Arthropoda  Insecta  Lepidoptera  Saturniidae  Lonomia   \n",
      "4  Costa_Rica  Arthropoda  Insecta  Lepidoptera  Saturniidae  Lonomia   \n",
      "\n",
      "                 Species  \n",
      "0  Lonomia_santarosensis  \n",
      "1  Lonomia_santarosensis  \n",
      "2  Lonomia_santarosensis  \n",
      "3  Lonomia_santarosensis  \n",
      "4  Lonomia_santarosensis  \n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "## Reading my Result"
  },
  {
   "cell_type": "code",
   "source": [
    "my_result = pd.read_csv(\"../../data/current/primer-finder-fixed3.csv\", sep=\";\")\n",
    "\n",
    "my_result['f_index'] = my_result['f_index'].astype(int)\n",
    "my_result['b_index'] = my_result['b_index'].astype(int)\n",
    "my_result['f_match'] = my_result['f_match'].astype(str)\n",
    "\n",
    "my_result['region'] = my_result.apply(lambda x: x['read'][(x['f_index'] + len(x['f_match'])):x['b_index']], axis=1)\n",
    "my_result['region_length'] = my_result['region'].str.len()\n",
    "print(my_result.head())"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:35.322922Z",
     "start_time": "2025-02-01T16:31:39.336421Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         BOLD ID       Read ID     Country            Phylum      Class  \\\n",
      "0  >BOLD:AAA7085  BLPDA1021-18  Costa_Rica  tax=p:Arthropoda  c:Insecta   \n",
      "1  >BOLD:AAA7085  MHMYN6354-14  Costa_Rica  tax=p:Arthropoda  c:Insecta   \n",
      "2  >BOLD:AAA7085   MHASB734-07  Costa_Rica  tax=p:Arthropoda  c:Insecta   \n",
      "3  >BOLD:AAA7085   MHATB310-06  Costa_Rica  tax=p:Arthropoda  c:Insecta   \n",
      "4  >BOLD:AAA7085  BLPAA6663-17  Costa_Rica  tax=p:Arthropoda  c:Insecta   \n",
      "\n",
      "           Order         Family      Genus                  Species  f_score  \\\n",
      "0  o:Lepidoptera  f:Saturniidae  g:Lonomia  s:Lonomia_santarosensis       49   \n",
      "1  o:Lepidoptera  f:Saturniidae  g:Lonomia  s:Lonomia_santarosensis       49   \n",
      "2  o:Lepidoptera  f:Saturniidae  g:Lonomia  s:Lonomia_santarosensis       49   \n",
      "3  o:Lepidoptera  f:Saturniidae  g:Lonomia  s:Lonomia_santarosensis       49   \n",
      "4  o:Lepidoptera  f:Saturniidae  g:Lonomia  s:Lonomia_santarosensis       49   \n",
      "\n",
      "                      f_match  f_index  b_score                  b_match  \\\n",
      "0  GGAACAGGATGAACGGTATACCCTCC      305       46  CCTGTTTTAGCAGGAGCTATTAC   \n",
      "1  GGAACAGGATGAACGGTATACCCTCC      319       46  CCTGTTTTAGCAGGAGCTATTAC   \n",
      "2  GGAACAGGATGAACGGTATACCCTCC      319       46  CCTGTTTTAGCAGGAGCTATTAC   \n",
      "3  GGAACAGGATGAACGGTATACCCTCC      319       46  CCTGTTTTAGCAGGAGCTATTAC   \n",
      "4  GGAACAGGATGAACGGTATACCCTCC      319       46  CCTGTTTTAGCAGGAGCTATTAC   \n",
      "\n",
      "   b_index                                               read  \\\n",
      "0      536  TCTTTGGAATTTGGGCAGGAATAGTAGGAACTTCTTTAAGTTTATT...   \n",
      "1      550  TACTTTATATTTTATCTTTGGAATTTGGGCAGGAATAGTAGGAACT...   \n",
      "2      550  TACTTTATATTTTATCTTTGGAATTTGGGCAGGAATAGTAGGAACT...   \n",
      "3      550  TACTTTATATTTTATCTTTGGAATTTGGGCAGGAATAGTAGGAACT...   \n",
      "4      550  TACTTTATATTTTATCTTTGGAATTTGGGCAGGAATAGTAGGAACT...   \n",
      "\n",
      "                                              region  region_length  \n",
      "0  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...            205  \n",
      "1  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...            205  \n",
      "2  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...            205  \n",
      "3  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...            205  \n",
      "4  TTTATCCTCTAACATTGCTCATGGAGGGTCTTCTGTAGATCTAGCT...            205  \n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## Compare Mine and Cutadapt\n",
    "* result of my perfect matches with cutadapt result\n",
    "* filter all which dont exist in cutadapt\n",
    "* what else?"
   ]
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### Are there any areas which differ for perfect matches?"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:37.429389Z",
     "start_time": "2025-02-01T16:32:35.367370Z"
    }
   },
   "cell_type": "code",
   "source": [
    "perfect_matches = my_result[~((my_result['f_score'] != 52) | (my_result['b_score'] != 46))]\n",
    "filtered_df = cutadapt_result[cutadapt_result['Read ID'].isin(perfect_matches['Read ID'])]"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:37.473758Z",
     "start_time": "2025-02-01T16:32:37.460386Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(f'Perfect matches: {perfect_matches.shape[0]}')\n",
    "print(f'Perfect matches found in cutadapt result: {filtered_df.shape[0]}')\n",
    "\n",
    "dif = perfect_matches.shape[0] - filtered_df.shape[0]\n",
    "print(f'Diff: {dif}, or {dif / perfect_matches.shape[0] * 100}% were not found in the cutadapt result.\\n')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Perfect matches: 898561\n",
      "Perfect matches found in cutadapt result: 898561\n",
      "Diff: 0, or 0.0% were not found in the cutadapt result.\n",
      "\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:39.322143Z",
     "start_time": "2025-02-01T16:32:37.518470Z"
    }
   },
   "cell_type": "code",
   "source": [
    "merged_df = filtered_df[['Read ID', 'Sequence']].merge(\n",
    "    perfect_matches[['Read ID', 'region']],\n",
    "    on='Read ID',\n",
    "    how='inner'\n",
    ").rename(columns={\n",
    "    'Sequence': 'ca_sequence',\n",
    "    'region': 'my_sequence'\n",
    "})\n",
    "different_sequences = merged_df[~(merged_df['ca_sequence'] == merged_df['my_sequence'])]\n",
    "print(different_sequences.shape[0])\n",
    "print(different_sequences.head())"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "28\n",
      "              Read ID                                        ca_sequence  \\\n",
      "153755  GBMNE23810-21  TTTATCTGGAGGTATTGCCCATGGGGGTGCTTCCGTAGATTTAGCT...   \n",
      "684885    NEPTA859-13  CTTATCAGCAAATATTGCTCATAGTGGTAGATCAGTTGATTTAGCA...   \n",
      "684898    NEPTA862-13  CTTATCAGCAAATATTGCTCATAGTGGTAGATCAGTTGATTTAGCA...   \n",
      "696620    GCOL7710-16  TTTATCTTCTAATATTACTCATAGAGGAGCTTCTGTTGATTTAGCT...   \n",
      "696621    GENHP169-11  TTTATCTTCTAATATTACTCATAGAGGAGCTTCTGTTGATTTAGCT...   \n",
      "\n",
      "                                              my_sequence  \n",
      "153755  TTTATCTGGAGGTATTGCCCATGGGGGTGCTTCCGTAGATTTAGCT...  \n",
      "684885  CTTATCAGCAAATATTGCTCATAGTGGTAGATCAGTTGATTTAGCA...  \n",
      "684898  CTTATCAGCAAATATTGCTCATAGTGGTAGATCAGTTGATTTAGCA...  \n",
      "696620  TTTATCTTCTAATATTACTCATAGAGGAGCTTCTGTTGATTTAGCT...  \n",
      "696621  TTTATCTTCTAATATTACTCATAGAGGAGCTTCTGTTGATTTAGCT...  \n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:39.466417Z",
     "start_time": "2025-02-01T16:32:39.452656Z"
    }
   },
   "cell_type": "code",
   "source": [
    "different_sequences['ca_length'] = different_sequences['ca_sequence'].str.len()\n",
    "different_sequences['my_length'] = different_sequences['my_sequence'].str.len()\n",
    "# 28 / 898561 * 100"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Julian\\AppData\\Local\\Temp\\ipykernel_2936\\3086744800.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  different_sequences['ca_length'] = different_sequences['ca_sequence'].str.len()\n",
      "C:\\Users\\Julian\\AppData\\Local\\Temp\\ipykernel_2936\\3086744800.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  different_sequences['my_length'] = different_sequences['my_sequence'].str.len()\n"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### why do those (28) entires fail with cutadapt?"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:39.717275Z",
     "start_time": "2025-02-01T16:32:39.485418Z"
    }
   },
   "cell_type": "code",
   "source": [
    "perfect_failed = my_result[my_result['Read ID'].isin(different_sequences['Read ID'])]\n",
    "perfect_failed = different_sequences.merge(\n",
    "    perfect_failed,\n",
    "    on='Read ID',\n",
    "    how='inner'\n",
    ").drop(columns=['region', 'region_length'])"
   ],
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T16:32:39.789879Z",
     "start_time": "2025-02-01T16:32:39.751367Z"
    }
   },
   "cell_type": "code",
   "source": "perfect_failed.to_csv('../../data/current/perfect_failed.csv', index=False)",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### whats up with my bad results, when one result was perfect??"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T18:15:04.789019Z",
     "start_time": "2025-02-01T18:14:49.150437Z"
    }
   },
   "cell_type": "code",
   "source": [
    "should_be_good = my_result[~((my_result['f_score'] != 52) & (my_result['b_score'] != 46))]\n",
    "filtered_df2 = cutadapt_result[cutadapt_result['Read ID'].isin(should_be_good['Read ID'])]\n",
    "merged_df2 = should_be_good.merge(\n",
    "    filtered_df2,\n",
    "    on='Read ID',\n",
    "    how='inner'\n",
    ").rename(columns={\n",
    "    'Sequence': 'ca_sequence',\n",
    "    'region': 'my_sequence'\n",
    "})\n",
    "\n",
    "merged_df2['ca_length'] = merged_df2['ca_sequence'].str.len()\n",
    "merged_df2['my_length'] = merged_df2['my_sequence'].str.len()"
   ],
   "outputs": [],
   "execution_count": 57
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T18:21:19.395987Z",
     "start_time": "2025-02-01T18:21:18.742074Z"
    }
   },
   "cell_type": "code",
   "source": [
    "merged_with_diff = merged_df2[merged_df2['ca_sequence'] != merged_df2['my_sequence']]\n",
    "#merged_df2_buckets = merged_with_diff.groupby(['b_score']).size().reset_index(name='Count')\n",
    "#merged_df2_buckets\n",
    "merged_with_diff = merged_with_diff[(merged_with_diff['my_length'] <= 185) & (merged_with_diff['my_length'] >= 140)].sort_values('my_length', ascending=False)\n",
    "merged_with_diff['remainder'] = merged_with_diff.apply(lambda x: len(x['read'][x['f_index']+231::]), axis=1)\n",
    "merged_with_diff.groupby(['remainder']).size().reset_index()\n"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "    remainder     0\n",
       "0           0   471\n",
       "1           1    23\n",
       "2           2     1\n",
       "3           3  1817\n",
       "4           4   213\n",
       "5           5   215\n",
       "6           6  5173\n",
       "7           7   164\n",
       "8           8   531\n",
       "9           9  2216\n",
       "10         10   205\n",
       "11         11   477\n",
       "12         12  1516\n",
       "13         13    24\n",
       "14         14    12\n",
       "15         15   307\n",
       "16         16     1\n",
       "17         22     1"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>remainder</th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>471</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>1817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>5173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>2216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12</td>\n",
       "      <td>1516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>13</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>14</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>15</td>\n",
       "      <td>307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 64
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T18:29:57.258836Z",
     "start_time": "2025-02-01T18:29:57.174836Z"
    }
   },
   "cell_type": "code",
   "source": "merged_df",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "              Read ID                                        ca_sequence  \\\n",
       "0         SPMNP492-22  ACTTTCATCTAATATTGCTCATAACGGATCTTCTGTTGATTTAGCT...   \n",
       "1         SPTMA398-07  TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...   \n",
       "2         INSPH096-19  TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...   \n",
       "3         SPTMA557-09  TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...   \n",
       "4         INSPH172-20  TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...   \n",
       "...               ...                                                ...   \n",
       "898556  GBMNA32630-19  TCTTTCATCCGGTATTGCCCATGCTGGTGGATCTGTAGATTTAGCT...   \n",
       "898557   LEPMY2719-22  TTTATCATCTAACATTGCTCATAGAGGAAGATCAGTAGATTTAGCT...   \n",
       "898558    LIMBC627-11  ACTTTCATCTAACATTGCCCATAGAGGAAGATCTGTTGACTTAGCA...   \n",
       "898559   SARBF2798-16  TCTCTCAGCTAACATTGCCCATAGCGGCTCTTCAGTTGACTTAGCC...   \n",
       "898560    GWOTD940-12  ACTTTCATCTAATATTGCTCATAGTGGAAGATCTGTAGATCTAGCT...   \n",
       "\n",
       "                                              my_sequence  \n",
       "0       ACTTTCATCTAATATTGCTCATAACGGATCTTCTGTTGATTTAGCT...  \n",
       "1       TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...  \n",
       "2       TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...  \n",
       "3       TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...  \n",
       "4       TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...  \n",
       "...                                                   ...  \n",
       "898556  TCTTTCATCCGGTATTGCCCATGCTGGTGGATCTGTAGATTTAGCT...  \n",
       "898557  TTTATCATCTAACATTGCTCATAGAGGAAGATCAGTAGATTTAGCT...  \n",
       "898558  ACTTTCATCTAACATTGCCCATAGAGGAAGATCTGTTGACTTAGCA...  \n",
       "898559  TCTCTCAGCTAACATTGCCCATAGCGGCTCTTCAGTTGACTTAGCC...  \n",
       "898560  ACTTTCATCTAATATTGCTCATAGTGGAAGATCTGTAGATCTAGCT...  \n",
       "\n",
       "[898561 rows x 3 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read ID</th>\n",
       "      <th>ca_sequence</th>\n",
       "      <th>my_sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SPMNP492-22</td>\n",
       "      <td>ACTTTCATCTAATATTGCTCATAACGGATCTTCTGTTGATTTAGCT...</td>\n",
       "      <td>ACTTTCATCTAATATTGCTCATAACGGATCTTCTGTTGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SPTMA398-07</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>INSPH096-19</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SPTMA557-09</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>INSPH172-20</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "      <td>TTTATCCTCAAATATTGCCCATAGTGGAAGATCTGTTGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898556</th>\n",
       "      <td>GBMNA32630-19</td>\n",
       "      <td>TCTTTCATCCGGTATTGCCCATGCTGGTGGATCTGTAGATTTAGCT...</td>\n",
       "      <td>TCTTTCATCCGGTATTGCCCATGCTGGTGGATCTGTAGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898557</th>\n",
       "      <td>LEPMY2719-22</td>\n",
       "      <td>TTTATCATCTAACATTGCTCATAGAGGAAGATCAGTAGATTTAGCT...</td>\n",
       "      <td>TTTATCATCTAACATTGCTCATAGAGGAAGATCAGTAGATTTAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898558</th>\n",
       "      <td>LIMBC627-11</td>\n",
       "      <td>ACTTTCATCTAACATTGCCCATAGAGGAAGATCTGTTGACTTAGCA...</td>\n",
       "      <td>ACTTTCATCTAACATTGCCCATAGAGGAAGATCTGTTGACTTAGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898559</th>\n",
       "      <td>SARBF2798-16</td>\n",
       "      <td>TCTCTCAGCTAACATTGCCCATAGCGGCTCTTCAGTTGACTTAGCC...</td>\n",
       "      <td>TCTCTCAGCTAACATTGCCCATAGCGGCTCTTCAGTTGACTTAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>898560</th>\n",
       "      <td>GWOTD940-12</td>\n",
       "      <td>ACTTTCATCTAATATTGCTCATAGTGGAAGATCTGTAGATCTAGCT...</td>\n",
       "      <td>ACTTTCATCTAATATTGCTCATAGTGGAAGATCTGTAGATCTAGCT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>898561 rows × 3 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 66
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "#### does cutadapt always cut at least 3 base pairs? - yes:"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T18:31:55.067306Z",
     "start_time": "2025-02-01T18:31:25.124502Z"
    }
   },
   "cell_type": "code",
   "source": [
    "merged_with_remainder = my_result.merge(\n",
    "    cutadapt_result,\n",
    "    on='Read ID',\n",
    "    how='inner'\n",
    ")\n",
    "merged_with_remainder['remainder'] = merged_with_remainder[merged_with_remainder['Sequence'] != ''].apply(\n",
    "    lambda x: len(x['read'].split(x['Sequence'])[-1]), axis=1\n",
    ")\n",
    "\n",
    "merged_with_remainder.groupby(['remainder']).size().reset_index(name='Count').sort_values(by=['remainder'], ascending=[True])"
   ],
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'read'",
     "output_type": "error",
     "traceback": [
      "\u001B[1;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[1;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3805\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[1;34m(self, key)\u001B[0m\n\u001B[0;32m   3804\u001B[0m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[1;32m-> 3805\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_engine\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mcasted_key\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   3806\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m \u001B[38;5;28;01mas\u001B[39;00m err:\n",
      "File \u001B[1;32mindex.pyx:167\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[1;34m()\u001B[0m\n",
      "File \u001B[1;32mindex.pyx:196\u001B[0m, in \u001B[0;36mpandas._libs.index.IndexEngine.get_loc\u001B[1;34m()\u001B[0m\n",
      "File \u001B[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7081\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[1;34m()\u001B[0m\n",
      "File \u001B[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7089\u001B[0m, in \u001B[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001B[1;34m()\u001B[0m\n",
      "\u001B[1;31mKeyError\u001B[0m: 'read'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001B[1;31mKeyError\u001B[0m                                  Traceback (most recent call last)",
      "Cell \u001B[1;32mIn[67], line 6\u001B[0m\n\u001B[0;32m      1\u001B[0m merged_with_remainder \u001B[38;5;241m=\u001B[39m my_result\u001B[38;5;241m.\u001B[39mmerge(\n\u001B[0;32m      2\u001B[0m     cutadapt_result,\n\u001B[0;32m      3\u001B[0m     on\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mRead ID\u001B[39m\u001B[38;5;124m'\u001B[39m,\n\u001B[0;32m      4\u001B[0m     how\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124minner\u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[0;32m      5\u001B[0m )\n\u001B[1;32m----> 6\u001B[0m merged_with_remainder[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mremainder\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m=\u001B[39m \u001B[43mmerged_df\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;28;43;01mlambda\u001B[39;49;00m\u001B[43m \u001B[49m\u001B[43mx\u001B[49m\u001B[43m:\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;28;43mlen\u001B[39;49m\u001B[43m(\u001B[49m\u001B[43mx\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mread\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43msplit\u001B[49m\u001B[43m(\u001B[49m\u001B[43mx\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mca_sequence\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[43m)\u001B[49m\u001B[43m)\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43maxis\u001B[49m\u001B[38;5;241;43m=\u001B[39;49m\u001B[38;5;241;43m1\u001B[39;49m\u001B[43m)\u001B[49m\n\u001B[0;32m      7\u001B[0m merged_with_diff\u001B[38;5;241m.\u001B[39mgroupby([\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mremainder\u001B[39m\u001B[38;5;124m'\u001B[39m])\u001B[38;5;241m.\u001B[39msize()\u001B[38;5;241m.\u001B[39mreset_index()\u001B[38;5;241m.\u001B[39msort_values(by\u001B[38;5;241m=\u001B[39m[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mCount\u001B[39m\u001B[38;5;124m'\u001B[39m], ascending\u001B[38;5;241m=\u001B[39m[\u001B[38;5;28;01mFalse\u001B[39;00m])\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\frame.py:10374\u001B[0m, in \u001B[0;36mDataFrame.apply\u001B[1;34m(self, func, axis, raw, result_type, args, by_row, engine, engine_kwargs, **kwargs)\u001B[0m\n\u001B[0;32m  10360\u001B[0m \u001B[38;5;28;01mfrom\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[38;5;21;01mpandas\u001B[39;00m\u001B[38;5;21;01m.\u001B[39;00m\u001B[38;5;21;01mcore\u001B[39;00m\u001B[38;5;21;01m.\u001B[39;00m\u001B[38;5;21;01mapply\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[38;5;28;01mimport\u001B[39;00m frame_apply\n\u001B[0;32m  10362\u001B[0m op \u001B[38;5;241m=\u001B[39m frame_apply(\n\u001B[0;32m  10363\u001B[0m     \u001B[38;5;28mself\u001B[39m,\n\u001B[0;32m  10364\u001B[0m     func\u001B[38;5;241m=\u001B[39mfunc,\n\u001B[1;32m   (...)\u001B[0m\n\u001B[0;32m  10372\u001B[0m     kwargs\u001B[38;5;241m=\u001B[39mkwargs,\n\u001B[0;32m  10373\u001B[0m )\n\u001B[1;32m> 10374\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43mop\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\u001B[38;5;241m.\u001B[39m__finalize__(\u001B[38;5;28mself\u001B[39m, method\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mapply\u001B[39m\u001B[38;5;124m\"\u001B[39m)\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\apply.py:916\u001B[0m, in \u001B[0;36mFrameApply.apply\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m    913\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mraw:\n\u001B[0;32m    914\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mapply_raw(engine\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mengine, engine_kwargs\u001B[38;5;241m=\u001B[39m\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mengine_kwargs)\n\u001B[1;32m--> 916\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply_standard\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\apply.py:1063\u001B[0m, in \u001B[0;36mFrameApply.apply_standard\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m   1061\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[38;5;21mapply_standard\u001B[39m(\u001B[38;5;28mself\u001B[39m):\n\u001B[0;32m   1062\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mengine \u001B[38;5;241m==\u001B[39m \u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mpython\u001B[39m\u001B[38;5;124m\"\u001B[39m:\n\u001B[1;32m-> 1063\u001B[0m         results, res_index \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mapply_series_generator\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1064\u001B[0m     \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[0;32m   1065\u001B[0m         results, res_index \u001B[38;5;241m=\u001B[39m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mapply_series_numba()\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\apply.py:1081\u001B[0m, in \u001B[0;36mFrameApply.apply_series_generator\u001B[1;34m(self)\u001B[0m\n\u001B[0;32m   1078\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m option_context(\u001B[38;5;124m\"\u001B[39m\u001B[38;5;124mmode.chained_assignment\u001B[39m\u001B[38;5;124m\"\u001B[39m, \u001B[38;5;28;01mNone\u001B[39;00m):\n\u001B[0;32m   1079\u001B[0m     \u001B[38;5;28;01mfor\u001B[39;00m i, v \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28menumerate\u001B[39m(series_gen):\n\u001B[0;32m   1080\u001B[0m         \u001B[38;5;66;03m# ignore SettingWithCopy here in case the user mutates\u001B[39;00m\n\u001B[1;32m-> 1081\u001B[0m         results[i] \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mfunc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mv\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43margs\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;241;43m*\u001B[39;49m\u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mkwargs\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1082\u001B[0m         \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(results[i], ABCSeries):\n\u001B[0;32m   1083\u001B[0m             \u001B[38;5;66;03m# If we have a view on v, we need to make a copy because\u001B[39;00m\n\u001B[0;32m   1084\u001B[0m             \u001B[38;5;66;03m#  series_generator will swap out the underlying data\u001B[39;00m\n\u001B[0;32m   1085\u001B[0m             results[i] \u001B[38;5;241m=\u001B[39m results[i]\u001B[38;5;241m.\u001B[39mcopy(deep\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mFalse\u001B[39;00m)\n",
      "Cell \u001B[1;32mIn[67], line 6\u001B[0m, in \u001B[0;36m<lambda>\u001B[1;34m(x)\u001B[0m\n\u001B[0;32m      1\u001B[0m merged_with_remainder \u001B[38;5;241m=\u001B[39m my_result\u001B[38;5;241m.\u001B[39mmerge(\n\u001B[0;32m      2\u001B[0m     cutadapt_result,\n\u001B[0;32m      3\u001B[0m     on\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mRead ID\u001B[39m\u001B[38;5;124m'\u001B[39m,\n\u001B[0;32m      4\u001B[0m     how\u001B[38;5;241m=\u001B[39m\u001B[38;5;124m'\u001B[39m\u001B[38;5;124minner\u001B[39m\u001B[38;5;124m'\u001B[39m\n\u001B[0;32m      5\u001B[0m )\n\u001B[1;32m----> 6\u001B[0m merged_with_remainder[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mremainder\u001B[39m\u001B[38;5;124m'\u001B[39m] \u001B[38;5;241m=\u001B[39m merged_df\u001B[38;5;241m.\u001B[39mapply(\u001B[38;5;28;01mlambda\u001B[39;00m x: \u001B[38;5;28mlen\u001B[39m(\u001B[43mx\u001B[49m\u001B[43m[\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mread\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m]\u001B[49m\u001B[38;5;241m.\u001B[39msplit(x[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mca_sequence\u001B[39m\u001B[38;5;124m'\u001B[39m])), axis\u001B[38;5;241m=\u001B[39m\u001B[38;5;241m1\u001B[39m)\n\u001B[0;32m      7\u001B[0m merged_with_diff\u001B[38;5;241m.\u001B[39mgroupby([\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mremainder\u001B[39m\u001B[38;5;124m'\u001B[39m])\u001B[38;5;241m.\u001B[39msize()\u001B[38;5;241m.\u001B[39mreset_index()\u001B[38;5;241m.\u001B[39msort_values(by\u001B[38;5;241m=\u001B[39m[\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mCount\u001B[39m\u001B[38;5;124m'\u001B[39m], ascending\u001B[38;5;241m=\u001B[39m[\u001B[38;5;28;01mFalse\u001B[39;00m])\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\series.py:1121\u001B[0m, in \u001B[0;36mSeries.__getitem__\u001B[1;34m(self, key)\u001B[0m\n\u001B[0;32m   1118\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_values[key]\n\u001B[0;32m   1120\u001B[0m \u001B[38;5;28;01melif\u001B[39;00m key_is_scalar:\n\u001B[1;32m-> 1121\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43m_get_value\u001B[49m\u001B[43m(\u001B[49m\u001B[43mkey\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1123\u001B[0m \u001B[38;5;66;03m# Convert generator to list before going through hashable part\u001B[39;00m\n\u001B[0;32m   1124\u001B[0m \u001B[38;5;66;03m# (We will iterate through the generator there to check for slices)\u001B[39;00m\n\u001B[0;32m   1125\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m is_iterator(key):\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\series.py:1237\u001B[0m, in \u001B[0;36mSeries._get_value\u001B[1;34m(self, label, takeable)\u001B[0m\n\u001B[0;32m   1234\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_values[label]\n\u001B[0;32m   1236\u001B[0m \u001B[38;5;66;03m# Similar to Index.get_value, but we do not fall back to positional\u001B[39;00m\n\u001B[1;32m-> 1237\u001B[0m loc \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mindex\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mget_loc\u001B[49m\u001B[43m(\u001B[49m\u001B[43mlabel\u001B[49m\u001B[43m)\u001B[49m\n\u001B[0;32m   1239\u001B[0m \u001B[38;5;28;01mif\u001B[39;00m is_integer(loc):\n\u001B[0;32m   1240\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_values[loc]\n",
      "File \u001B[1;32mZ:\\Programs\\mambaforge\\envs\\MA\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3812\u001B[0m, in \u001B[0;36mIndex.get_loc\u001B[1;34m(self, key)\u001B[0m\n\u001B[0;32m   3807\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(casted_key, \u001B[38;5;28mslice\u001B[39m) \u001B[38;5;129;01mor\u001B[39;00m (\n\u001B[0;32m   3808\u001B[0m         \u001B[38;5;28misinstance\u001B[39m(casted_key, abc\u001B[38;5;241m.\u001B[39mIterable)\n\u001B[0;32m   3809\u001B[0m         \u001B[38;5;129;01mand\u001B[39;00m \u001B[38;5;28many\u001B[39m(\u001B[38;5;28misinstance\u001B[39m(x, \u001B[38;5;28mslice\u001B[39m) \u001B[38;5;28;01mfor\u001B[39;00m x \u001B[38;5;129;01min\u001B[39;00m casted_key)\n\u001B[0;32m   3810\u001B[0m     ):\n\u001B[0;32m   3811\u001B[0m         \u001B[38;5;28;01mraise\u001B[39;00m InvalidIndexError(key)\n\u001B[1;32m-> 3812\u001B[0m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mKeyError\u001B[39;00m(key) \u001B[38;5;28;01mfrom\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[38;5;21;01merr\u001B[39;00m\n\u001B[0;32m   3813\u001B[0m \u001B[38;5;28;01mexcept\u001B[39;00m \u001B[38;5;167;01mTypeError\u001B[39;00m:\n\u001B[0;32m   3814\u001B[0m     \u001B[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001B[39;00m\n\u001B[0;32m   3815\u001B[0m     \u001B[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001B[39;00m\n\u001B[0;32m   3816\u001B[0m     \u001B[38;5;66;03m#  the TypeError.\u001B[39;00m\n\u001B[0;32m   3817\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_check_indexing_error(key)\n",
      "\u001B[1;31mKeyError\u001B[0m: 'read'"
     ]
    }
   ],
   "execution_count": 67
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "#### what is the best score, if the b-primer is outside the read?\n",
    "oof, ok, there are regions with len 190 lol"
   ]
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T20:16:28.103471Z",
     "start_time": "2025-02-01T20:16:01.075360Z"
    }
   },
   "cell_type": "code",
   "source": [
    "outside = my_result\n",
    "outside['remainder'] = outside.apply(lambda x: len(x['read'][x['f_index']+231::]), axis=1)\n",
    "outside = outside[outside['remainder'] <= 12]\n"
   ],
   "outputs": [],
   "execution_count": 83
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T20:26:12.189513Z",
     "start_time": "2025-02-01T20:26:12.146684Z"
    }
   },
   "cell_type": "code",
   "source": "outside_no_region = outside[outside['region'] == ''].copy()\n",
   "outputs": [],
   "execution_count": 98
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T20:26:14.326027Z",
     "start_time": "2025-02-01T20:26:14.238516Z"
    }
   },
   "cell_type": "code",
   "source": [
    "outside_no_region['region'] = outside_no_region.apply(lambda x: x['read'][(x['f_index'] + len(x['f_match'])):x['b_index']], axis=1)\n",
    "outside_no_region"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "               BOLD ID        Read ID               Country            Phylum  \\\n",
       "16593    >BOLD:AAC9112    BBCNP186-13                Canada  tax=p:Arthropoda   \n",
       "19620    >BOLD:AAB6956    VVGAP469-20               Ukraine  tax=p:Arthropoda   \n",
       "22005    >BOLD:AAM1247    MAHEM349-10              Pakistan  tax=p:Arthropoda   \n",
       "22368    >BOLD:AAM1247    MAHEM382-10              Pakistan  tax=p:Arthropoda   \n",
       "22481    >BOLD:AAM1247    MAHEM370-10              Pakistan  tax=p:Arthropoda   \n",
       "...                ...            ...                   ...               ...   \n",
       "2378502  >BOLD:ADC8118   GBMTG4732-16                   NaN  tax=p:Arthropoda   \n",
       "2379746  >BOLD:ADR7996     DISA506-19  South_Atlantic_Ocean  tax=p:Arthropoda   \n",
       "2380340  >BOLD:ACL6603    ZSMDB022-14                 Chile  tax=p:Arthropoda   \n",
       "2380872  >BOLD:AAJ5917    GBLN0644-06                   NaN  tax=p:Arthropoda   \n",
       "2380886  >BOLD:AAN2237  GBMNF14322-22                   NaN  tax=p:Arthropoda   \n",
       "\n",
       "                  Class          Order            Family           Genus  \\\n",
       "16593       c:Arachnida      o:Araneae     f:Linyphiidae  g:Bathyphantes   \n",
       "19620         c:Insecta   o:Coleoptera   f:Curculionidae       g:Larinus   \n",
       "22005         c:Insecta    o:Hemiptera     f:Aleyrodidae       g:Bemisia   \n",
       "22368         c:Insecta    o:Hemiptera     f:Aleyrodidae       g:Bemisia   \n",
       "22481         c:Insecta    o:Hemiptera     f:Aleyrodidae       g:Bemisia   \n",
       "...                 ...            ...               ...             ...   \n",
       "2378502     c:Arachnida      o:Ixodida       f:Argasidae         g:Argas   \n",
       "2379746  c:Malacostraca      o:Isopoda  f:Sphaeromatidae  g:Exosphaeroma   \n",
       "2380340       c:Insecta   o:Coleoptera      f:Dytiscidae       g:Rhantus   \n",
       "2380872       c:Insecta  o:Lepidoptera     f:Nymphalidae        g:Aemona   \n",
       "2380886       c:Insecta    o:Hemiptera        f:Coreidae        g:Mozena   \n",
       "\n",
       "                               Species  f_score  \\\n",
       "16593          s:Bathyphantes_pallidus       32   \n",
       "19620                s:Larinus_minutus       31   \n",
       "22005                 s:Bemisia_tabaci       29   \n",
       "22368                 s:Bemisia_tabaci       29   \n",
       "22481                 s:Bemisia_tabaci       29   \n",
       "...                                ...      ...   \n",
       "2378502               s:Argas_miniatus       29   \n",
       "2379746  s:Exosphaeroma_truncatitelson       30   \n",
       "2380340             s:Rhantus_signatus       30   \n",
       "2380872                  s:Aemona_lena       31   \n",
       "2380886                s:Mozena_lunata       30   \n",
       "\n",
       "                                 f_match  f_index  b_score  \\\n",
       "16593                GGAGCTGGTTGGACTGTAT      319       27   \n",
       "19620        GTACTTTGACTGTATA-TGTATATCCA      552       27   \n",
       "22005         GGTATCTCGTCTATTCAGATTATCCT      589       28   \n",
       "22368         GGTATCTCGTCTATTCAGATTATCCT      589       28   \n",
       "22481         GGTATCTCGTCTATTCAGATTATCCT      589       28   \n",
       "...                                  ...      ...      ...   \n",
       "2378502     GG---TGGTTCGGAACAAGTTATATCGT      782       30   \n",
       "2379746      GAATTGAAATGACCTGTACAAACCCTC      434       29   \n",
       "2380340   GGTACGCAGGAAATAACTGATTTAATGCCT      603       33   \n",
       "2380872  GGGTCAAGGAT---CTGGTATATCTACCACC      215       28   \n",
       "2380886         GGAACATTAGGAATAATTTACGCA      408       43   \n",
       "\n",
       "                              b_match  b_index  \\\n",
       "16593           CCTATTTTAATTGGGGGTTTT      166   \n",
       "19620      CCACTCCTGGCTGGGTCGTTCAATTA       26   \n",
       "22005          CTTATTTTACCAGGCTGGTATT       19   \n",
       "22368          CTTATTTTACCAGGCTGGTATT       19   \n",
       "22481          CTTATTTTACCAGGCTGGTATT       19   \n",
       "...                               ...      ...   \n",
       "2378502      CCGTGATTTATCAGGGAG-AATCA      588   \n",
       "2379746  CCCTGATATTCAAGCCGCAG-TATCCAC       38   \n",
       "2380340       CATGTATTAGCTCTAG-AATTAC      559   \n",
       "2380872   CCGAGTTTTCGATCTAGGTGACTTTTA       31   \n",
       "2380886       CCTGTGCTAGCAGGAGCAATCAC      201   \n",
       "\n",
       "                                                      read region  \\\n",
       "16593    GACGTTATATTTAATTTTTGGTGCCTGGGCAGCTATGGTAGGGACT...          \n",
       "19620    TGCGGTCCTCGGATTCTGTTCCCGGACCACTCCTGGCTGAGGGTCG...          \n",
       "22005    TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...          \n",
       "22368    TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...          \n",
       "22481    TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...          \n",
       "...                                                    ...    ...   \n",
       "2378502  ATTTCTTTAATAAGATATATATTATTGCTTGTTATAGTGTTGATTA...          \n",
       "2379746  AAAACATCGTCGGTGTCTACTTTCGCCGATAAGATCTGCCCAGTGA...          \n",
       "2380340  CACCCGGCAGGACGTCTCGCGGGCGTGCCAGTTTAACACCGCGGGC...          \n",
       "2380872  CTGCACTGTGAAGACGTGTTGGATGCGTCTGCCGAGTTTTCGATCT...          \n",
       "2380886  TCAAGCAATTTATCACACAGAGGAGCATCCGTTGACTTAGCAATTT...          \n",
       "\n",
       "         region_length  remainder  \n",
       "16593                0          0  \n",
       "19620                0          0  \n",
       "22005                0          0  \n",
       "22368                0          9  \n",
       "22481                0         10  \n",
       "...                ...        ...  \n",
       "2378502              0          0  \n",
       "2379746              0          0  \n",
       "2380340              0          0  \n",
       "2380872              0          0  \n",
       "2380886              0          0  \n",
       "\n",
       "[4651 rows x 19 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BOLD ID</th>\n",
       "      <th>Read ID</th>\n",
       "      <th>Country</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>f_score</th>\n",
       "      <th>f_match</th>\n",
       "      <th>f_index</th>\n",
       "      <th>b_score</th>\n",
       "      <th>b_match</th>\n",
       "      <th>b_index</th>\n",
       "      <th>read</th>\n",
       "      <th>region</th>\n",
       "      <th>region_length</th>\n",
       "      <th>remainder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16593</th>\n",
       "      <td>&gt;BOLD:AAC9112</td>\n",
       "      <td>BBCNP186-13</td>\n",
       "      <td>Canada</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Arachnida</td>\n",
       "      <td>o:Araneae</td>\n",
       "      <td>f:Linyphiidae</td>\n",
       "      <td>g:Bathyphantes</td>\n",
       "      <td>s:Bathyphantes_pallidus</td>\n",
       "      <td>32</td>\n",
       "      <td>GGAGCTGGTTGGACTGTAT</td>\n",
       "      <td>319</td>\n",
       "      <td>27</td>\n",
       "      <td>CCTATTTTAATTGGGGGTTTT</td>\n",
       "      <td>166</td>\n",
       "      <td>GACGTTATATTTAATTTTTGGTGCCTGGGCAGCTATGGTAGGGACT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19620</th>\n",
       "      <td>&gt;BOLD:AAB6956</td>\n",
       "      <td>VVGAP469-20</td>\n",
       "      <td>Ukraine</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Coleoptera</td>\n",
       "      <td>f:Curculionidae</td>\n",
       "      <td>g:Larinus</td>\n",
       "      <td>s:Larinus_minutus</td>\n",
       "      <td>31</td>\n",
       "      <td>GTACTTTGACTGTATA-TGTATATCCA</td>\n",
       "      <td>552</td>\n",
       "      <td>27</td>\n",
       "      <td>CCACTCCTGGCTGGGTCGTTCAATTA</td>\n",
       "      <td>26</td>\n",
       "      <td>TGCGGTCCTCGGATTCTGTTCCCGGACCACTCCTGGCTGAGGGTCG...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22005</th>\n",
       "      <td>&gt;BOLD:AAM1247</td>\n",
       "      <td>MAHEM349-10</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Hemiptera</td>\n",
       "      <td>f:Aleyrodidae</td>\n",
       "      <td>g:Bemisia</td>\n",
       "      <td>s:Bemisia_tabaci</td>\n",
       "      <td>29</td>\n",
       "      <td>GGTATCTCGTCTATTCAGATTATCCT</td>\n",
       "      <td>589</td>\n",
       "      <td>28</td>\n",
       "      <td>CTTATTTTACCAGGCTGGTATT</td>\n",
       "      <td>19</td>\n",
       "      <td>TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22368</th>\n",
       "      <td>&gt;BOLD:AAM1247</td>\n",
       "      <td>MAHEM382-10</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Hemiptera</td>\n",
       "      <td>f:Aleyrodidae</td>\n",
       "      <td>g:Bemisia</td>\n",
       "      <td>s:Bemisia_tabaci</td>\n",
       "      <td>29</td>\n",
       "      <td>GGTATCTCGTCTATTCAGATTATCCT</td>\n",
       "      <td>589</td>\n",
       "      <td>28</td>\n",
       "      <td>CTTATTTTACCAGGCTGGTATT</td>\n",
       "      <td>19</td>\n",
       "      <td>TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22481</th>\n",
       "      <td>&gt;BOLD:AAM1247</td>\n",
       "      <td>MAHEM370-10</td>\n",
       "      <td>Pakistan</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Hemiptera</td>\n",
       "      <td>f:Aleyrodidae</td>\n",
       "      <td>g:Bemisia</td>\n",
       "      <td>s:Bemisia_tabaci</td>\n",
       "      <td>29</td>\n",
       "      <td>GGTATCTCGTCTATTCAGATTATCCT</td>\n",
       "      <td>589</td>\n",
       "      <td>28</td>\n",
       "      <td>CTTATTTTACCAGGCTGGTATT</td>\n",
       "      <td>19</td>\n",
       "      <td>TCATCCAGAAGTTTATGTTCTTATTTTACCAGGCTTTGGTATTGTT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378502</th>\n",
       "      <td>&gt;BOLD:ADC8118</td>\n",
       "      <td>GBMTG4732-16</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Arachnida</td>\n",
       "      <td>o:Ixodida</td>\n",
       "      <td>f:Argasidae</td>\n",
       "      <td>g:Argas</td>\n",
       "      <td>s:Argas_miniatus</td>\n",
       "      <td>29</td>\n",
       "      <td>GG---TGGTTCGGAACAAGTTATATCGT</td>\n",
       "      <td>782</td>\n",
       "      <td>30</td>\n",
       "      <td>CCGTGATTTATCAGGGAG-AATCA</td>\n",
       "      <td>588</td>\n",
       "      <td>ATTTCTTTAATAAGATATATATTATTGCTTGTTATAGTGTTGATTA...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2379746</th>\n",
       "      <td>&gt;BOLD:ADR7996</td>\n",
       "      <td>DISA506-19</td>\n",
       "      <td>South_Atlantic_Ocean</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Malacostraca</td>\n",
       "      <td>o:Isopoda</td>\n",
       "      <td>f:Sphaeromatidae</td>\n",
       "      <td>g:Exosphaeroma</td>\n",
       "      <td>s:Exosphaeroma_truncatitelson</td>\n",
       "      <td>30</td>\n",
       "      <td>GAATTGAAATGACCTGTACAAACCCTC</td>\n",
       "      <td>434</td>\n",
       "      <td>29</td>\n",
       "      <td>CCCTGATATTCAAGCCGCAG-TATCCAC</td>\n",
       "      <td>38</td>\n",
       "      <td>AAAACATCGTCGGTGTCTACTTTCGCCGATAAGATCTGCCCAGTGA...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380340</th>\n",
       "      <td>&gt;BOLD:ACL6603</td>\n",
       "      <td>ZSMDB022-14</td>\n",
       "      <td>Chile</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Coleoptera</td>\n",
       "      <td>f:Dytiscidae</td>\n",
       "      <td>g:Rhantus</td>\n",
       "      <td>s:Rhantus_signatus</td>\n",
       "      <td>30</td>\n",
       "      <td>GGTACGCAGGAAATAACTGATTTAATGCCT</td>\n",
       "      <td>603</td>\n",
       "      <td>33</td>\n",
       "      <td>CATGTATTAGCTCTAG-AATTAC</td>\n",
       "      <td>559</td>\n",
       "      <td>CACCCGGCAGGACGTCTCGCGGGCGTGCCAGTTTAACACCGCGGGC...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380872</th>\n",
       "      <td>&gt;BOLD:AAJ5917</td>\n",
       "      <td>GBLN0644-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Lepidoptera</td>\n",
       "      <td>f:Nymphalidae</td>\n",
       "      <td>g:Aemona</td>\n",
       "      <td>s:Aemona_lena</td>\n",
       "      <td>31</td>\n",
       "      <td>GGGTCAAGGAT---CTGGTATATCTACCACC</td>\n",
       "      <td>215</td>\n",
       "      <td>28</td>\n",
       "      <td>CCGAGTTTTCGATCTAGGTGACTTTTA</td>\n",
       "      <td>31</td>\n",
       "      <td>CTGCACTGTGAAGACGTGTTGGATGCGTCTGCCGAGTTTTCGATCT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2380886</th>\n",
       "      <td>&gt;BOLD:AAN2237</td>\n",
       "      <td>GBMNF14322-22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>tax=p:Arthropoda</td>\n",
       "      <td>c:Insecta</td>\n",
       "      <td>o:Hemiptera</td>\n",
       "      <td>f:Coreidae</td>\n",
       "      <td>g:Mozena</td>\n",
       "      <td>s:Mozena_lunata</td>\n",
       "      <td>30</td>\n",
       "      <td>GGAACATTAGGAATAATTTACGCA</td>\n",
       "      <td>408</td>\n",
       "      <td>43</td>\n",
       "      <td>CCTGTGCTAGCAGGAGCAATCAC</td>\n",
       "      <td>201</td>\n",
       "      <td>TCAAGCAATTTATCACACAGAGGAGCATCCGTTGACTTAGCAATTT...</td>\n",
       "      <td></td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4651 rows × 19 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 99
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T20:29:31.415145Z",
     "start_time": "2025-02-01T20:29:31.248145Z"
    }
   },
   "cell_type": "code",
   "source": "cutadapt_result[cutadapt_result['Read ID'] == \"ANICJ009-10\"]",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Sequence, BOLD ID, Read ID, Country, Phylum, Class, Order, Family, Genus, Species]\n",
       "Index: []"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>BOLD ID</th>\n",
       "      <th>Read ID</th>\n",
       "      <th>Country</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 101
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
